### DISCUSSION TITLE
Normal mode in file and buffer picker

### DISCUSSION THREAD

**Discussion by txtyash** [helix-editor/helix#10599] [Ideas]
Add normal mode in file picker or buffer picker & also make it the default
mode in buffer picker. This way navigation will be easier with j,k and in
buffer picker it is more likely to be a faster way.

**Comment by EpocSquadron** [comment]
This starts to get tricky to implement, and breaks the pattern of using readline like bindings when in prompts, as is default in most shells. On the other hand, there's been discussion of removing a lot of the readline/emacs bindings in insert mode to encourage modal mindset. I think I'd be in support of this if it were implemented well.

**Comment by joegm** [comment]
Most of the big advantages normal mode has are to do with multiple lines and multiple cursors. Normal mode for single line command prompt seems overkill as you're practically never going to need it, so wouldn't be significantly faster in reality.

j/k to scroll the menus does sound nice, if people think it would be worth implementing for that alone. We could even go farther and have bindings like `d` in the buffer picker to close a buffer, stuff like that, but of course that would likely require a fairly big rework.

**Comment by EpocSquadron** [comment]
There is some precedent as zsh's readline implementation supports setting to vi-mode instead of emacs. It does however start in insert mode and require you to press escape to get to normal. You're right that a lot of the normal mode power in helix wouldn't carry over, it makes more sense when its vim you're implementing. Some subset might still be worth looking into though.

**Comment by rsHalford** [comment]
My go to for insert mode line scrolling is to always bind to Ctrl-j/k. Using Ctrl-p/n in vim has never seemed to click for me. However, I imagine this would just be something to wait on with picker mode remapping?

**Comment by the-mikedavis** [comment]
Yep that would more likely fall under https://github.com/helix-editor/helix/issues/2757

**Comment by premell** [comment]
I find it almost a bit claustrophobic that you cant select the text in the file picker. Quite often i write something I want to search for, then I want to copy the same string. However that is not possible in the current picker. On the other hand, it is nice that the picker closes with just one q press instead of two.

I think I would prefer normal mode though

**Comment by jfaz13** [comment]
Honestly I think something that would work is adding shortcuts to open files in the buffer picker faster. Right now they are numbered, so you can do e.g. `5 + enter` and it would work, but it would be more expedient if you could just press 5 and it opens it automatically. Once you get past 9 it becomes an issue (`1` vs `13`), but we can just mix and match numbers and letters for ones after 9 (`1a`, `1b`, etc). I guess a disadvantage of doing this is that you don't preview the file before opening it, but maybe that can be chorded away, or maybe if a key beforehand is pressed it'll preview before opening. Maybe instead of listing with numbers they could just start with letters for easy accessibility. Just a thought.

**Comment by shved** [comment]
Ctrl + n and Ctrl + p to navigate up and down seem to be exhaustive enough for all it ought to be.

**Comment by Cooksey99** [comment]
I recently made the following PR (#14634) and stumbled upon this discussion. I feel like my solution is quite simple, without being intrusive to users who may not have the same preference. There are likely some tweaks that should be made, but I wanted to get some feedback before adding too much.
